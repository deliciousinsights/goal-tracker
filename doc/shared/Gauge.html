<!DOCTYPE html><html lang="en"><head><title>shared/Gauge</title></head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0"><meta name="groc-relative-root" content="../"><meta name="groc-document-path" content="shared/Gauge"><meta name="groc-project-path" content="src/shared/Gauge.js"><link rel="stylesheet" type="text/css" media="all" href="../assets/style.css"><script type="text/javascript" src="../assets/behavior.js"></script><body><div id="meta"><div class="file-path">src/shared/Gauge.js</div></div><div id="document"><div class="segment"><div class="comments "><div class="wrapper"><h1 id="jauge">Jauge</h1>
<p>Réutilisé absolument partout dans l’application…  Un parfait exemple de
composant réutilisable, du coup !</p></div></div><div class="code"><div class="wrapper"><span class="hljs-keyword">import</span> LinearProgress <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/material/LinearProgress'</span>

<span class="hljs-keyword">import</span> { nonNegativeInteger, positiveInteger } <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/prop-types'</span>

export default function Gauge({ max, value }) {
  return (
    &lt;LinearProgress
      style={{ height: 8 }}
      variant=<span class="hljs-string">'determinate'</span>
      value={normalize(value, max)}
    /&gt;
  )
}</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>En définissant les valeurs par défaut dans <code>defaultProps</code> au lieu des valeurs
par défaut de la signature, on s’assure qu’elles seront bien prises en compte
avant l’exploitation des <code>propTypes</code> ci-après.</p></div></div><div class="code"><div class="wrapper">Gauge.defaultProps = {
  max: <span class="hljs-number">100</span>,
}

Gauge.propTypes = {
  max: positiveInteger,
  value: nonNegativeInteger.isRequired,
}

<span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">normalize</span>(<span class="hljs-params">value, max</span>) </span>{
  <span class="hljs-keyword">return</span> value * (<span class="hljs-number">100</span> / max)
}</div></div></div></div></body></html>