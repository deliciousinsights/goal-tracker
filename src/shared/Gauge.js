// Jauge
// =====
//
// Réutilisé absolument partout dans l’application…  Un parfait exemple de
// composant réutilisable, du coup !

import LinearProgress from '@mui/material/LinearProgress'

import { nonNegativeInteger, positiveInteger } from '../shared/prop-types'

export default function Gauge({ max, value }) {
  return (
    <LinearProgress
      style={{ height: 8 }}
      variant='determinate'
      value={normalize(value, max)}
    />
  )
}

// En définissant les valeurs par défaut dans `defaultProps` au lieu des valeurs
// par défaut de la signature, on s’assure qu’elles seront bien prises en compte
// avant l’exploitation des `propTypes` ci-après.
Gauge.defaultProps = {
  max: 100,
}

Gauge.propTypes = {
  max: positiveInteger,
  value: nonNegativeInteger.isRequired,
}

function normalize(value, max) {
  return value * (100 / max)
}
