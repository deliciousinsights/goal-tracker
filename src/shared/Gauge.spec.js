// Jauge (tests)
// =============

import { render, screen } from '@testing-library/react'

import Gauge from './Gauge'

describe('<Gauge />', () => {
  it('should render appropriately', () => {
    render(<Gauge value={50} />)
    // On récupère le composant dont le rôle est une barre de progression
    // (défini en interne par le `<LinearProgress/>` au sein de `<Gauge/>`).
    const progressBar = screen.getByRole('progressbar')
    // On vérifie qu’il est bien en mode “determinate” (la barre ne cycle pas)
    expect(progressBar).toHaveClass('MuiLinearProgress-determinate')
    // On vérifie que notre style personnalisé est appliqué.
    expect(progressBar).toHaveStyle({ height: '8px' })
    // On vérifie la valeur normalisée
    expect(progressBar).toHaveAttribute('aria-valuenow', '50')
  })

  it('should normalize value on custom max', () => {
    render(<Gauge value={20} max={80} />)
    // On vérifie la valeur normalisée
    expect(screen.getByRole('progressbar')).toHaveAttribute(
      'aria-valuenow',
      '25'
    )
  })

  it('should otherwise match the expected snapshots', () => {
    const { container } = render(<Gauge value={50} />)

    // On se contente de comparer le rendu à un
    // [*snapshot*](https://jestjs.io/docs/en/snapshot-testing) pris en
    // développement quand le composant fonctionnait comme prévu.
    expect(container).toMatchSnapshot()
  })
})
