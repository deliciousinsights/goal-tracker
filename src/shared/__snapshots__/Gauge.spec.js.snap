// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gauge /> should otherwise match the expected snapshots 1`] = `
<div>
  <span
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
    role="progressbar"
    style="height: 8px;"
  >
    <span
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
      style="transform: translateX(-50%);"
    />
  </span>
</div>
`;
